eventstore/query_jsonmode_test.go

```diff
@@ -0,0 +1,658 @@
+package eventstore
+
+import (
+	"sort"
+	"strings"
+	"testing"
+)
+
+const (
+	testQueryJsonModeSchema = `{
+		"$schema": "http://json-schema.org/draft-04/schema#",
+		"$ref": "#/definitions/book",
+		"definitions": {
+		   "book": {
+			  "required": [
+				 "ID",
+				 "Title",
+				 "Author",
+				 "Meta"
+			  ],
+			  "properties": {
+				 "Author": {
+					"type": "string"
+				 },
+				 "Banned": {
+					"type": "boolean"
+				 },
+				 "ID": {
+					"type": "string"
+				 },
+				 "Meta": {
+					"$schema": "http://json-schema.org/draft-04/schema#",
+					"$ref": "#/definitions/bookStats"
+				 },
+				 "Title": {
+					"type": "string"
+				 }
+			  },
+			  "additionalProperties": false,
+			  "type": "object"
+		   },
+		   "bookStats": {
+			  "required": [
+				 "TotalReads",
+				 "Rating"
+			  ],
+			  "properties": {
+				 "Rating": {
+					"type": "number"
+				 },
+				 "TotalReads": {
+					"type": "integer"
+				 }
+			  },
+			  "additionalProperties": false,
+			  "type": "object"
+		   }
+		}
+	 }`
+)
+
+type jsonQueryTest struct {
+	name    string
+	query   JSONQuery
+	resIdx  []int
+	ordered bool
+}
+
+var (
+	jsonSampleData = []string{
+		`{"ID": "", "Title": "Title1", "Banned": true,  "Author": "Author1", "Meta": { "TotalReads": 100, "Rating": 3.2 }}`,
+		`{"ID": "", "Title": "Title2", "Banned": false, "Author": "Author1", "Meta": { "TotalReads": 150, "Rating": 4.1 }}`,
+		`{"ID": "", "Title": "Title3", "Banned": false, "Author": "Author2", "Meta": { "TotalReads": 120, "Rating": 4.6 }}`,
+		`{"ID": "", "Title": "Title4", "Banned": true,  "Author": "Author3", "Meta": { "TotalReads": 1000, "Rating": 2.6 }}`,
+	}
+
+	boolTrue   = true
+	boolFalse  = false
+	title      = "Title"
+	title0     = "Title1"
+	title1     = "Title2"
+	title3     = "Title3"
+	titleMax   = "Title5000"
+	totreadEq1 = 100.0
+	totreadMid = 124.36
+	totreadEq2 = 150.0
+	totreadMax = 1221.25
+	totreadMin = 12.22
+	ratingMin  = 1.0
+	rating1    = 3.2
+	ratingMid  = 3.6
+	ratingMax  = 5.0
+
+	jsonQueries = []jsonQueryTest{
```

Comment:
A bunch of tests to good coverage, since all this feels brittle:
- Ands for each possible datatype `string`, `bool`, `float`
- Ors
- Ordering (again for each possible datatype to be sure, with asc and desc)
- Mixing ones: `and` with `or` with `sort` (asc/desc)

Also, helps to understand a bit how this is used, capabilities, etc.